ContinousBenchmarking.md

```diff
@@ -0,0 +1,175 @@
+
+
+## Goals and Non-Goals
+*This version of the proposal will focus on a minimally viable system. These goals will evolve once the initial goal is accomplished and we decide to bring in more complexity.
+#### Goals:
+- Update benchmarking scripts so they can be started in a consistent format and run to completion without manual intervention.
+- Focus on getting system up and running on a single machine.
+- Design a simple system for deploying the results page.
+- Allow garage developers to detect regressions in algorithm performance caused by code changes
+
+#### Non-Goals: 
+- Worry about scalability too early.  MVP is top priority.
+- Design for frugality over simplicity
+- Manage web servers
+
+## Design Overview
+*The input and output interfaces of the benchmarking scripts must be normalized (start commands, and results).
+This initial design will use a cron job script ran on a single local machine. On a nightly basis the script shall:
+1. pull the latest garage code. 
+2. Run commands to launch benchmarking jobs
+3. Detect when the jobs are completed
+4. Run script to generate results page with plots.
+5. Deploy results.html to Github pages or similar.
+
+### Options:
+
+#### Python
+Perhaps we can just write a single Python script to accomplish steps 1 - 3? This way we can import the benchmarking scripts as modules, and run it like that.  
+
+#### Shell 
+We need to launch the conda env after pulling the latest garage.  Maybe this workflow would be more appropriate as a shell script.  
+
+For step 4, I am thinking we can just have a html/js template, and use grunt/gruntfile to do some simple replacements of data.
+
+For step 5, There seems to be many options, I'd appreciate all suggestions!
+
+
+
+## Detailed Design
+Rough outline of script:
+    
+    #pull latest garage -- something like this.
+    git clone https://github.com/rlworkgroup/garage.git 
+
+    #export conda
+    CONDA_SH="${HOME}/miniconda2/etc/profile.d/conda.sh"
+    . "${CONDA_SH}"
+
+    #activate conda
+    conda activate garage
+
+    #work around until GLEW issue is resolved.
+    export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libGLEW.so:/usr/lib/nvidia-390/libGL.so
+
+    #run benchmark
+    nose2 tests.benchmarks.test_benchmark_ddpg
+    
+    echo "done with benchmarks"
+    
+    #copy the results graphs, and json into the display's resources folder.
+    rm -rf garage-dashboard/resources/*
+    cp -r benchmark_ddpg/latest/*.png garage-dashboard/resources
+    cp benchmark_ddpg/latest/results.json garage-dashboard/resources
+    
+    #conda deactivate
+    conda deactivate
+
+
+Display:
+
+To collect the results, I propose we modify each benchmark.py scripts to output a single JSON.  
+    
+     result  = {
+                    "time" : "November 11th 1:24:00",
+                    "branch":"Master",
+                    "graphs":[
+                        "./resources/HalfCheetah-v2_benchmark.png",
```

Comment:
Can you keep around the raw numerical results? This way if we change the frontend we can reprocess into new plots.